client.go

```diff
@@ -925,12 +936,42 @@ func anyError(resp *http.Response, err error) error {
 }
 
 // doRequest creates and performs an http request.
-func (c *Client) doRequest(host *URI, method, path string, headers map[string]string, reader io.Reader) (*http.Response, error) {
-	req, err := makeRequest(host, method, path, headers, reader)
+func (c *Client) doRequest(host *URI, method, path string, headers map[string]string, data []byte) (*http.Response, error) {
+	var resp *http.Response
+	var err error
+	var req *http.Request
+	tries := 1 + c.retries
+	sleepTime := c.minRetrySleepTime
+	for tries > 0 {
+		req, err = makeRequest(host, method, path, headers, data)
+		if err != nil {
+			return nil, errors.Wrap(err, "building request")
+		}
+		tries--
+		resp, err = c.client.Do(req)
+		if err == nil {
+			if resp.StatusCode >= 200 && resp.StatusCode < 300 {
+				return resp, nil
+			}
+			// Pilosa nodes sometimes return 400, we retry in that case.
+			// No need to retry in other 4xx cases.
+			if resp.StatusCode > 400 && resp.StatusCode < 500 {
+				return resp, nil
+			}
+			c.logger.Printf("request failed with: %d, retrying (%d)", resp.StatusCode, tries)
+		} else {
+			c.logger.Printf("request failed with: %s, retrying (%d)", err.Error(), tries)
+		}
+		time.Sleep(sleepTime)
+		sleepTime *= 2
+		if sleepTime > c.maxRetrySleepTime {
+			sleepTime = c.maxRetrySleepTime
+		}
+	}
 	if err != nil {
-		return nil, errors.Wrap(err, "building request")
+		return nil, errors.Wrap(err, "max retries exceeded")
```

Comment:
if the `err` is nil down here, then we'll fail with "max retries exceeded" and no information about what problem was actually occurring. It seems like this could happen in the expected case of a 400 or 5XX when the cluster is in state STARTING. Similarly, the log message above gives the status code rather than an actual error message which would be more informative.

Probably the easiest thing would be to see if the err from the request is nil, and if it is, but the status is 400 or 5XX to set err to a new error which includes the message from the response. Then use a single logger line, and a single return line down here since the error should be guaranteed to be non-nil.

Reply:
updated.